shared/reducers/fs.js

```diff
@@ -14,29 +14,58 @@ export default function(state: Types.State = initialState, action: FsGen.Actions
     case FsGen.filePreviewLoaded:
       return state.update('pathItems', metas => metas.set(action.payload.path, action.payload.meta))
     case FsGen.folderListLoaded: {
-      const toMerge = action.payload.pathItems.filter((item, path) => {
-        if (item.type !== 'folder') {
-          return true
-        }
+      const toMerge = action.payload.pathItems.map((item, path) => {
+        if (item.type !== 'folder') return item
         const original = state.pathItems.get(path)
-        if (original && original.progress === 'loaded' && item.progress === 'pending') {
+        if (!original || original.type !== 'folder') return item
+        if (original.progress === 'loaded' && item.progress !== 'pending') {
           // Don't override a loaded item into pending. This is specifically
           // for the case where user goes back out of a folder where we could
           // override the folder into an empty one. With this, next user
           // navigates into the folder they would see the old list (instead of
           // placeholder), which then gets updated when we hear back from RPC.
-          return false
+          return original
         }
-        return true
+        // Since both `folderListLoaded` and `favoritesLoaded` can change
+        // `pathItems`, we need to make sure that neither one clobbers the
+        // other's work.
+        return item
+          .set('badgeCount', original.badgeCount)
+          .set('tlfMeta', original.tlfMeta)
+          .set('favoriteChildren', original.favoriteChildren)
       })
-      return state
+      const s = state
         .mergeIn(['pathItems'], toMerge)
         .update('loadingPaths', loadingPaths => loadingPaths.delete(action.payload.path))
+      return s
     }
     case FsGen.folderListLoad:
       return state.update('loadingPaths', loadingPaths => loadingPaths.add(action.payload.path))
+    case FsGen.favoritesLoad:
+      return state
+    case FsGen.favoritesLoaded:
+      const toMerge = action.payload.folders.mapEntries(([path, item]) => {
+        // We ForceType because Flow keeps thinking this is a _PathItem not a FolderPathItem.
+        const original: $ForceType = state.pathItems.get(path) || Constants.makeFolder({name: item.name})
+        // This cannot happen, but it's needed to make Flow happy.
+        if (original.type !== 'folder') return [path, original]
+
+        return [
+          path,
+          // Since both `folderListLoaded` and `favoritesLoaded` can change
+          // `pathItems`, we need to make sure that neither one clobbers the
+          // other's work.
```

Comment:
I tried it like that first, and it was a nightmare to get working. Since the sorting and other stuff works on path items, I'd have to pigeonhole `FavoriteItem`s into `PathItem`s at some point anyway. I guess I could have a separate map and just have them both typed `PathItem`?

Reply:
Additionally: eventually this is going to go away since we're planning on putting this functionality into `SimpleFS`. The current approach will result in fewer changes to complete that refactor.

Reply:
Yeah I'm inclined to stick with the current approach the more I think about it, to avoid having to build custom code into the presentation layer.

Reply:
OK sounds good; thanks for the context!